galley/pkg/config/analysis/analyzers/deployment/services.go

```diff
@@ -0,0 +1,128 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package deployment
+
+import (
+	apps_v1 "k8s.io/api/apps/v1"
+	core_v1 "k8s.io/api/core/v1"
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/galley/pkg/config/meta/metadata"
+	"istio.io/istio/galley/pkg/config/meta/schema/collection"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+type ServiceAssociationAnalyzer struct{}
+
+var _ analysis.Analyzer = &ServiceAssociationAnalyzer{}
+
+type PortMap map[int32]ProtocolMap
+type ProtocolMap map[core_v1.Protocol][]ServiceSpecWithName
+type ServiceNames []string
+type ServiceSpecWithName struct {
+	Name string
+	Spec *core_v1.ServiceSpec
+}
+
+func (s *ServiceAssociationAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "deployment.MultiServiceAnalyzer",
+		Inputs: collection.Names{
+			metadata.K8SCoreV1Services,
+			metadata.K8SAppsV1Deployments,
+		},
+	}
+}
+func (s *ServiceAssociationAnalyzer) Analyze(c analysis.Context) {
+	portMap := servicePortMap(c)
+
+	c.ForEach(metadata.K8SAppsV1Deployments, func(r *resource.Entry) bool {
+		s.analyzeDeployment(r, c, portMap)
+		return true
+	})
+}
+
+// servicePortMap build a map of ports and protocols for each Service. e.g. m[80]["TCP"] -> svcA, svcB, svcC
+func servicePortMap(c analysis.Context) PortMap {
+	portMap := PortMap{}
+
+	c.ForEach(metadata.K8SCoreV1Services, func(rs *resource.Entry) bool {
+		svc := rs.Item.(*core_v1.ServiceSpec)
+
+		for _, sPort := range svc.Ports {
+			// If it is the first occurrence of this port, create a ProtocolMap
+			if _, ok := portMap[sPort.Port]; !ok {
+				portMap[sPort.Port] = ProtocolMap{}
+			}
+
+			// Default protocol is TCP
+			protocol := sPort.Protocol
+			if protocol == "" {
+				protocol = core_v1.ProtocolTCP
+			}
+
+			// Appending the service information for the Port/Procotol combination
+			portMap[sPort.Port][protocol] = append(portMap[sPort.Port][protocol], ServiceSpecWithName{
+				rs.Metadata.Name.String(),
+				svc,
+			})
+		}
+
+		return true
+	})
+
+	return portMap
+}
+
+// analyzeDeployment analyzes the specific deployment given the service port map
+func (s *ServiceAssociationAnalyzer) analyzeDeployment(r *resource.Entry, c analysis.Context, pm PortMap) {
+	d := r.Item.(*apps_v1.Deployment)
+	for _, cont := range d.Spec.Template.Spec.Containers {
+		for _, portSt := range cont.Ports {
+			port := portSt.ContainerPort
```

Comment:
Istio used to require container port, but has not since 1.3

> I think comparing the exposed port here (if provided) with the port defined on the service could make sense as a separate analyzer, though... if you are specifying containerPort and associating a service, I think it's reasonable to warn if they don't line up.

This is not istio specific, so I would recommend against it. Although if you want to make a `kubectl analyze` that would be cool...

> For checking for service association conflicts, we should be able to just look at the pods associated with a service by checking the labels in the deployment template against the label selectors on the service, and making sure that for the services associated with a given pod don't have any conflicts.

Agree, just associate based on labels

Reply:
> > I think comparing the exposed port here (if provided) with the port defined on the service could make sense as a separate analyzer, though... if you are specifying containerPort and associating a service, I think it's reasonable to warn if they don't line up.
> 
> This is not istio specific, so I would recommend against it. Although if you want to make a `kubectl analyze` that would be cool...

That's true, but given that `kubectl analyze` doesn't exist and it's pretty easy to add here, maybe it's worth adding anyway? But only if enough users actually run into this problem to justify it, so I'm fine punting on that question until there's demand.

Reply:
I think there is two components here.

One is should istioctl handle non-istio specific things. I don't think so, but others do and we already do have parts of istioctl that do this

The other is should this specifically be an analyze, and I don't think so, since containerPort has no correlation to service port. You can put whatever you want in either one, and they aren't expected to be 1:1. Its not an error for them not to align

Reply:
@sushicw I changed the approach for the analyzer. I based the previous solution on deployments port, which it was wrong.
The approach on the latest push is through the selector/label matching and then applying the service/port/protocol combination.
Thanks for the patience :)

Reply:
btw, kiali has a warning for the mismatch between service and port. Someone explicitly asked it.
Documentation: [here](https://kiali.io/documentation/validations/#_deployment_exposing_same_port_as_service_not_found)